### PR TITLE
core: add zombie thread state

### PR DESCRIPTION
**Pull request by JulianHolzwarth**
### Contribution description

This pull request adds a new feature.
Zombie thread state is added (`STATUS_ZOMBIE`).
This is a thread state to keep the `thread_t` (holds thread's context data) and to keep the pid (make sure no other thread can have the pid) of a thread that wants/needs to keep it for another thread. Zombie threads cannot be started again, will never run and will only be removed when the function `thread_kill_zombie` is called. A thread has to call `thread_zombify` to make itself a zombie. The thread will then stop to run and change it thread state to `STATUS_ZOMBIE`. This is useful for posix pthreads.

This pr makes ps able to correctly display zombie threads.
This pr adds a test.

### Testing procedure

Run the test. 
The test shows ps output to make sure it is correctly displayed.

run: `BOARD=native make -C tests/thread_zombie/  flash test`
expected output:

```
...
main(): This is RIOT! (Version: 2020.01-devel-174-g6dd10-pr/thread_status_t/zombie)
start threads = 2
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | sleeping _ |   6 |  20480 (  420) | 0x56624120 | 0x56628f90 
	    | SUM                  |            |     |  49152 ( 3412)

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | sleeping _ |   6 |  20480 (  420) | 0x56624120 | 0x56628f90 
	  4 | nr3                  | sleeping _ |   6 |  20480 (  420) | 0x56629120 | 0x5662df90 
	    | SUM                  |            |     |  69632 ( 3832)

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | sleeping _ |   6 |  20480 (  420) | 0x56624120 | 0x56628f90 
	  4 | nr3                  | sleeping _ |   6 |  20480 (  420) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | sleeping _ |   6 |  20480 (  420) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  90112 ( 4252)

threads = 5
OK
other thread

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | zombie   _ |   6 |  20480 (  768) | 0x56624120 | 0x56628f90 
	  4 | nr3                  | sleeping _ |   6 |  20480 (  420) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | sleeping _ |   6 |  20480 (  420) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  90112 ( 4600)

threads = 5
OK
other thread

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | zombie   _ |   6 |  20480 (  768) | 0x56624120 | 0x56628f90 
	  4 | nr3                  | zombie   _ |   6 |  20480 (  768) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | sleeping _ |   6 |  20480 (  420) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  90112 ( 4948)

threads = 5
OK
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  4 | nr3                  | zombie   _ |   6 |  20480 (  768) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | sleeping _ |   6 |  20480 (  420) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  69632 ( 4180)

threads = 4
OK
other thread

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  4 | nr3                  | zombie   _ |   6 |  20480 (  768) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | zombie   _ |   6 |  20480 (  768) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  69632 ( 4528)

OK
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56638420 | 0x56638420
	  1 | idle                 | pending  Q |  15 |   8192 (  420) | 0x56636240 | 0x566380b0 
	  2 | main                 | running  Q |   7 |  12288 ( 2572) | 0x56633240 | 0x566360b0 
	  3 | nr2                  | sleeping _ |   6 |  20480 (  420) | 0x56624120 | 0x56628f90 
	  4 | nr3                  | zombie   _ |   6 |  20480 (  768) | 0x56629120 | 0x5662df90 
	  5 | nr4                  | zombie   _ |   6 |  20480 (  768) | 0x5662e120 | 0x56632f90 
	    | SUM                  |            |     |  90112 ( 4948)

OK

```

### CODE REVIEW

**Comment by waehlisch** [comment]
@kaspar030, i'm a bit confused: you add a block without giving technical reasons. can you elaborate? thanks.

**Comment by bergzand** [comment]
I think @kaspar030 is blocking because there are issues with the proposed solution for the problem. From how I interpret his comment, he would like the author to investigate an alternative solution that doesn't increase the size of the core RIOT threading system but is only implemented using the pthread module from the posix module. In terms of labels I guess he is blocking as it fails [review step 1](https://github.com/RIOT-OS/RIOT/blob/master/MAINTAINING.md#1---review-the-fundamentals) for him.

**Comment by kaspar030** [comment]
> can you elaborate? thanks.

I thought I was clear: why do we need this? 

"This is useful for posix pthreads." is not enough justification to add to RIOT's core.

"zombie threads" are a kludge in posix systems to get the exit code of a finished child process into the parent process by calling the `wait()` system call. RIOT does not have that kind of relationship between processes, so a zombie state is not needed for RIOT threads.

This should be contained within the pthreads implementation. (I just checked, apparently zombie state is already handled there).

> This pr makes ps able to correctly display zombie threads.

Is this the main motivation for this PR? Why not:

1. refactor `pthread_self()` into two functions, one returning the pthread_t (or NULL) for any given pid, and another `pthread_self() { return new_function(sched_active_pid);`

2. in ps, check each thread if it is a pthread using the new function from above, and if it is, print the thread state accordingly

No need to add anything "zombie thread" related to core.

**Comment by JulianHolzwarth** [comment]
> This pr makes ps able to correctly display zombie threads.

This was added for two reasons. First, to make testing easier. Second, making zombies explicit prevents ambiguous output when executing `ps`. If you do not want to have the `ps` change in there I will remove it.


Regarding the main question:
> why do we need this?

I need a way to keep the `thread_t` (holds thread's context data) of a thread and keep the pid of a thread after it died. So that no other thread can be created with the same pid after the thread died but before the other thread was able to run.
The posix thread was an example. It allows us both to have one `pthread_pid == pid` and
to be able to let a thread know when another thread is done.

> I just checked, apparently zombie state is already handled there

They are handled by missusing `msg_send_int @brief Send message from interrupt.`, and by creating another thread. This results into overhead. This PR introduces a solution that prevents this overhead. Why? Because the stack can be freed by the thread that calls join. That means it saves a thread (i.e., stack for memory, no need to create a thread, no need to run the thread ...).

To summarize:

1. I do not think that this feature is only beneficial for pthreads. For example: a thread sends a message and another thread is created which gets the same pid as the one that died. Then, the thread that got the message does not know the other thread died.
2. And it helps pthreads for sure.

**Comment by JulianHolzwarth** [comment]
Before PR:
```
$ LANG=C.UTF-8 BOARD=native make RIOT_CI_BUILD=1 -C tests/thread_basic/ flash
...
Building application "tests_thread_basic" for "native" with MCU "native".
   text	   data	    bss	    dec	    hex	filename
  23096	    584	  59940	  83620	  146a4	.../thread_basic/bin/native/tests_thread_basic.elf
true
```


After PR:
```
$ LANG=C.UTF-8 BOARD=native make RIOT_CI_BUILD=1 -C tests/thread_basic/ flash
...
Building application "tests_thread_basic" for "native" with MCU "native".

   text	   data	    bss	    dec	    hex	filename
  23096	    584	  59940	  83620	  146a4	...thread_basic/bin/native/tests_thread_basic.elf
true
```

**Comment by kaspar030** [comment]
> They are handled by missusing `msg_send_int @brief Send message from interrupt.`, and by creating another thread. This results into overhead. 

Yeah, the implementation looks a little brave.

> This PR introduces a solution that prevents this overhead. Why? Because the stack can be freed by the thread that calls join. That means it saves a thread (i.e., stack for memory, no need to create a thread, no need to run the thread ...).

I don't think a new thread state is needed. Just set a thread's pthread state to "PTS_ZOMBIE", the thread's own state to "STATUS_STOPPED", and yield.

In the joining thread, disable irqs, handle what needs handling, free the stack, then set `sched_threads[pid] = NULL`.

Well, that's what this PR's extra functions are doing (apart from the new state).
Let me think about it a bit, ok?

>     1. I do not think that this feature is only beneficial for pthreads. For example: a thread sends a message and another thread is created which gets the same pid as the one that died. Then, the thread that got the message does not know the other thread died.

I don't think this would be prevented with zombed threads (only for the arbitrary time it takes to call `join()`).

**Comment by JulianHolzwarth** [comment]
> I don't think a new thread state is needed. Just set a thread's pthread state to "PTS_ZOMBIE", the thread's own state to "STATUS_STOPPED", and yield.

I don't think we should use `STATUS_STOPPED` for it. Because a thread that has the status `STATUS_STOPPED` should be removed and a zombie thread not.
Meaning the status has a different purpose.

> 
> In the joining thread, disable irqs, handle what needs handling, free the stack, then set sched_threads[pid] = NULL.
> 
> Well, that's what this PR's extra functions are doing (apart from the new state).

The joining thread in `sys/posix/pthread/pthread.c` function `join` at the moment only handles posix thread handling (and yes disable irqs should be done but is not done yet) and will after this PR also free the stack. 
But `sched_threads[pid] = NULL` should not be done in posix because it modifies a core code(structure). In my opinion posix should not modify core code(structure). For this reason I want to merge this PR.

This PR's extra function kill/terminate the thread from another thread. (similar to kill but only for zombies)
To kill a thread you almost the same code as: `NORETURN void sched_task_exit(void)` (changed a little so it works from another thread)

The other function just sets the status. 

What do you think? @kaspar030

**Comment by kaspar030** [comment]
> What do you think? @kaspar030

ok, let's do it.

File: core/thread.c [hunk 1]
```diff
@@ -55,6 +55,47 @@ const char *thread_getname(kernel_pid_t pid)
...
+    if (irq_is_in()) {
+        return;
+    }
+
+    unsigned state = irq_disable();
+    sched_set_status((thread_t *)sched_active_thread, STATUS_ZOMBIE);
+    irq_restore(state);
+    thread_yield_higher();
+
+    sched_task_exit();
```
Comment (kaspar030) [inline core/thread.c]: This line should never be reached, right? The stack might not be valid anymore.

Should we maybe replace it with sth like `__builtin_unreachable()`?

At least there needs to be a comment.

File: core/thread.c [hunk 2]
```diff
@@ -55,6 +55,47 @@ const char *thread_getname(kernel_pid_t pid)
...
+        DEBUG("thread_kill: Thread does not exist!\n");
+    }
+    else if (thread->status == STATUS_ZOMBIE) {
+        DEBUG("thread_kill: Thread is a zombie.\n");
+
+        sched_threads[pid] = NULL;
+        sched_num_threads--;
+        sched_set_status(thread, STATUS_STOPPED); 
+
+        irq_restore(state);
```
Comment (kaspar030) [inline core/thread.c]: how about:

```
unsigned state = ...
int res = (int)STATUS_NOT_FOUND;

...
if () {
...
}
else if (thread->status == STATUS_ZOMBIE) {
  ...
  res = 0;
}
else {
 ...
}
irq_restore(state);
return res;
```

- saves one "irq_restore()"
- changes return to "0 on success, <0 on error"

File: tests/thread_zombie/main.c [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    else {
+        if (start_thread_count + 3 != current_thread_count) {
+            puts("Error wrong thread count");
+        }
+        else {
+            puts("OK");
+        }
+
+    }
+
```
Comment (kaspar030) [inline tests/thread_zombie/main.c]: extra newlines

File: core/thread.c (hunk 1 above)
Comment (kaspar030) [reply core/thread.c]: or `UNRACHABLE()`?

File: core/thread.c (hunk 2 above)
Comment (kaspar030) [reply core/thread.c]: > changes return to "0 on success, <0 on error"

scratch that, seems like other `thread_*` functions also return 1 on success.

Comment (kaspar030) [reply core/thread.c]: >     * saves one "irq_restore()"

we can do this as followup.

File: core/thread.c (hunk 1 above)
Comment (JulianHolzwarth) [reply core/thread.c]: yes, line 69: `sched_task_exit();` should never be reached.

I think it is a good idea to change it to `UNREACHABLE()` /  `__builtin_unreachable()` . To do that we have to change the function to a non-void function. (at least on my maschine)

**Comment by JulianHolzwarth** [comment]
@kaspar030 I pushed the commits. What do you think about them?

File: core/thread.c [hunk 4]
```diff
@@ -55,25 +55,28 @@ const char *thread_getname(kernel_pid_t pid)
 #endif
 }
 
-void thread_zombify(void)
+int thread_zombify(void)
```
Comment (kaspar030) [inline core/thread.c]: Does `NORETURN` instead of `int` work?

**Comment by kaspar030** [comment]
> @kaspar030 I pushed the commits. What do you think about them?

looking good, just the `int` bugs me, it shouldn't be necessary.

File: core/thread.c (hunk 4 above)
Comment (JulianHolzwarth) [reply core/thread.c]: It has to return when `irq_is_in()` is true.

**Comment by JulianHolzwarth** [comment]
@kaspar030 I agree.

File: core/thread.c (hunk 1 above)
Comment (kaspar030) [reply core/thread.c]: > To do that we have to change the function to a non-void function. (at least on my maschine)

What compiler / architecture is that?

**Comment by kaspar030** [comment]
I tried this:

```
#include <stdio.h>
void foo(void) {
    __builtin_unreachable();
}

int main(void)
{
    puts("Hello World!");
    foo();
...
}
```

... for a couple of targets, there's no issue. What's special on your machine? :wink:

**Comment by JulianHolzwarth** [comment]
@kaspar030 like this it does work( the thing you did ). But not the when it is in `thread_zombify`

**Comment by JulianHolzwarth** [comment]
I changed it back to void. Can you try removing ` thread_yield_higher();` and running the test.

**Comment by waehlisch** [comment]
@kaspar030 any time  to comment on @JulianHolzwarth fast replies?

**Comment by kaspar030** [comment]
> I changed it back to void. Can you try removing ` thread_yield_higher();` and running the test.

So it works now with using void?

**Comment by kaspar030** [comment]
> @kaspar030 any time to comment on @JulianHolzwarth fast replies?

Sorry, I was assuming @JulianHolzwarth was fixing the remaining CI errors. The tests still fail on some platforms.

File: core/include/thread.h [hunk 5]
```diff
@@ -398,6 +398,22 @@ void thread_yield(void);
  */
 void thread_yield_higher(void);
 
+/**
+ * @brief   Puts the current thread into zombie mode. Has to be terminated externally.
```
Comment (kaspar030) [inline core/include/thread.h]: Feel free to elaborate a bit on what that means.
Something along the line:

```
A thread in zombie state will never be scheduled again, but its scheduler entry and stack will be kept.
It is supposed to be cleaned up by @ref thread_kill_zombie().
```

Comment (JulianHolzwarth) [reply core/include/thread.h]: Yes, the thread becomes a zombie thread (will never be  scheduled again, but its scheduler entry and stack will be kept ).
And is supposed to be cleaned up by @ref thread_kill_zombie() (run from different thread/not the zombie thread because the zombie should never be able to run again)

**Comment by JulianHolzwarth** [comment]
travis fails because the commit messages are to long. I will change them when I rebase (I will also squash all commits except::
`core/include/sched.h: new thread_status_t thread state`
`core/thread.c: new function for zombie state`
`sys/ps/ps.c: ps can now handle zombie state threads`
`tests/thread_zombie/: new test for zombie state threads`

**Comment by JulianHolzwarth** [comment]
> > I changed it back to void. Can you try removing ` thread_yield_higher();` and running the test.
> 
> So it works now with using void?

I tried it on debian and Ubuntu and it does not work with `UNREACHABLE()` in `thread_zombify()`. (I removed  `thread_yield_higher()` and ran the test)

**Comment by JulianHolzwarth** [comment]
@kaspar030 Can I rebase on master? I think the `test_utils_interactive_sync` change makes it fail on Murdock.

**Comment by kaspar030** [comment]
perfekt!

**Comment by kaspar030** [comment]
> I tried it on debian and Ubuntu and it does not work with `UNREACHABLE()` in `thread_zombify()`. (I removed `thread_yield_higher()` and ran the test)

Ah, now I get it. You tested if `UNREACHABLE()` somehow ensures that the code cannot be reached. I think that's not the case. But IMO we can be perfectly sure about that, as we set the thread state to a value that means "not on run queue", so the scheduler will definitely not schedule the thread again.

**Comment by kaspar030** [comment]
please rebase and squash!

**Comment by JulianHolzwarth** [comment]
@kaspar030 I rebased. Also I changed the description of `thread_zombify` to:
```
/**
 * @brief   Puts the current thread into zombie state.
 * 
 * @details Does nothing when in ISR.
 *          A thread in zombie state will never be scheduled again,
 *          but its scheduler entry and stack will be kept.
 *          A zombie state thread is supposed to be cleaned up 
 *          by @ref thread_kill_zombie().
 */
void thread_zombify(void);
```

**Comment by JulianHolzwarth** [comment]
whitespace fix

**Comment by JulianHolzwarth** [comment]
murdock fails because:
`Pull request needs squashing according to its labels set on GitHub`

**Comment by waehlisch** [comment]
All comments addressed. Review labels set. Murdock is happy. We are done with that, right? Any reason to keep the block, @kaspar030?

**Comment by kaspar030** [comment]
ACK.
